### PR TITLE
greg/NFSE-5141-fix-perfc-get-cycles

### PR DESCRIPTION
**Pull request by beckerg**
Signed-off-by: Greg Becker <gbecker@micron.com>

### CODE REVIEW

**Comment by beckerg** [comment]
This patch leverages our jiffy clock to improve the accuracy of the time stamp counter frequency over time.  Normally it's pretty close to the bogomips values, but on some machines it can be many millions of hertz different, which can throw off measurements made in comparison to reading the clock (as is what was happening with the perfc_get_cycles() unit test.

File: lib/util/include/hse_util/platform.h [hunk 1]
```diff
@@ -83,31 +83,21 @@ hse_meminfo(unsigned long *freep, unsigned long *availp, unsigned int shift);
...
+extern volatile unsigned int hse_tsc_mult;
 
 static HSE_ALWAYS_INLINE u64
 cycles_to_nsecs(u64 cycles)
 {
-    if (HSE_UNLIKELY(cycles >= HSE_TSC_HWM))
-        return (cycles >> HSE_TSC_SHIFT) * hse_tsc_mult;
-
-    return (cycles * hse_tsc_mult) >> HSE_TSC_SHIFT;
+    return ((__uint128_t)cycles * hse_tsc_mult) >> HSE_TSC_SHIFT;
```
Comment (tristan957) [inline lib/util/include/hse_util/platform.h:100]: Not sure if this is the first use uint128_t...Is uint128_t guaranteed to be supported on all platforms? If not, I would like to bail out in the meson.build file.

This reminds me that we should error the build if aligned (?) and packed attributes aren't supported by the compiler.

If you don't feel like doing it, I can do both in the next PR.

**Review by tristan957 — APPROVED** [review]

**Comment by tristan957** [comment]
Thanks for fixing the unit tests.

File: lib/util/include/hse_util/platform.h (hunk 1 above)
Comment (beckerg) [reply lib/util/include/hse_util/platform.h:100]: I think technically "unsigned int128_t" is the correct form, but I used __int128_t on s390x and both gcc and clang accept this form.  Feel free to figure out what's best.  I think if the machine doesn't support 128-bit types the compiler just writes magic code to do it with whatever it has available.
I can't imagine any compiler used to build our stuff that didn't support aligned/packed, so should't hurt to test and fail if that's the case..

**Pull request merged**